src/plugins/flow.js

```diff
@@ -1053,6 +1062,68 @@ export default function (instance) {
     };
   });
 
+  // We need to support type parameter declarations for arrow functions. This
+  // is tricky. There are three situations we need to handle
+  //
+  // 1. This is either JSX or an arrow function. We'll try JSX first. If that
+  //    fails, we'll try an arrow function. If that fails, we'll throw the JSX
+  //    error.
+  // 2. This is an arrow function. We'll parse the type parameter declaration,
+  //    parse the rest, make sure the rest is an arrow function, and go from
+  //    there
+  // 3. This is neither. Just call the inner function
+  instance.extend("parseMaybeAssign", function (inner) {
+    return function (...args) {
+      let jsxError = null;
+      let state = this.state.clone();
```

Comment:
This isn't super obvious but cloning state is actually super expensive, for example there's a `tokens` array on there that gets cloned and this can sometimes number in the tens of thousands for really large files.